File: notifications/macosx.nim
# Copyright (C) Dominik Picheta. All rights reserved.
# MIT License. Look at license.txt for more info.
import os, asyncdispatch
{.passL: "-framework Foundation".}
{.passL: "-framework AppKit".}
{.compile: "macosx.m".}

# Types defined in the Objective C file.

type
  AdditionalButton = object
    title: cstring
    identifier: cstring

  NSApplication = distinct pointer
  ActivationInfo {.bycopy.} = object
    activationType: cint
    selectedActionTitle: cstring
    selectedActionIdentifier: cstring
    reply: cstring

  NotificationCallback = proc (info: ActivationInfo, data: pointer) {.cdecl.}
  NotificationState {.bycopy.} = object
    app: NSApplication
    onNotificationClick: NotificationCallback
    data: pointer

# Nim types
type
  NotificationCenterObj* = object
    state: NotificationState
    pollInterval: float # In ms
    pollTimeout: float # In ms
    polling: bool
    pollFut: Future[void]
    onNotificationClick: proc (info: ClickInfo) {.closure.}
  NotificationCenter* = ref NotificationCenterObj

  ClickKind* {.pure.} = enum
    None, ContentsClicked, ActionButtonClicked, Replied, AdditionalActionClicked
  ## Information about what the user clicked on the notification.
  ClickInfo* = object ## Apple call this an "Activation"
    case kind*: ClickKind
    of ClickKind.None, ClickKind.ContentsClicked,
       ClickKind.ActionButtonClicked: discard
    of ClickKind.Replied:
      message*: string
    of ClickKind.AdditionalActionClicked:
      selectedTitle*: string
      selectedIdentifier*: string

  NotificationError* = object of OSError

# Wrapped C functions

proc showNotification(title, subtitle, message, actionButtonTitle,
    otherButtonTitle: cstring, hasReplyButton: bool,
    additionalButtons: ptr AdditionalButton, additionalButtonsSize: cint,
    errorCode: ptr cint): cint {.importc, cdecl.}

proc createApp(onNotificationClick: NotificationCallback,
    data: pointer): NotificationState
    {.importc, cdecl.}

proc poll(app: NotificationState, timeout: cfloat) {.importc, cdecl.}

proc freeActivationInfo(info: ActivationInfo) {.importc, cdecl.}

# Nim procedures

proc newNotificationCenter*(
    onNotificationClick: proc (info: ClickInfo) {.closure.} = nil,
    pollInterval = 100.0,
    pollTimeout = 100.0): NotificationCenter =
  ## Creates a new NotificationCenter.
  ##
  ## The ``pollInterval`` parameter determines (in milliseconds) how often the
  ## Cocoa events should be read.
  ##
  ## The ``pollTimeout`` parameter determines (in milliseconds) how long the
  ## the notification center should wait for new Cocoa events.
  proc objCCallback(info: ActivationInfo, data: pointer) {.cdecl.} =
    let center = cast[ptr NotificationCenterObj](data)
    var clickKind = ClickKind.None
    case info.activationType
    of 0: clickKind = ClickKind.None
    of 1: clickKind = ClickKind.ContentsClicked
    of 2: clickKind = ClickKind.ActionButtonClicked
    of 3: clickKind = ClickKind.Replied
    of 4: clickKind = ClickKind.AdditionalActionClicked
    else:
      raise newException(NotificationError,
          "Unknown activation type, got " & $info.activationType)

    # The kinds above are actually not reported correctly.
    if not info.selectedActionTitle.isNil:
      clickKind = ClickKind.AdditionalActionClicked

    if not center.onNotificationClick.isNil:
      case clickKind
      of ClickKind.AdditionalActionClicked:
        center.onNotificationClick(ClickInfo(
          kind: clickKind,
          selectedTitle: $info.selectedActionTitle,
          selectedIdentifier: $info.selectedActionIdentifier
        ))
      of ClickKind.Replied:
        center.onNotificationClick(ClickInfo(
          kind: clickKind,
          message: $info.reply
        ))
      else:
        center.onNotificationClick(ClickInfo(
          kind: clickKind
        ))

    # Free the ActivationInfo struct's fields.
    freeActivationInfo(info)
    center.polling = false

  var ret: NotificationCenter
  new ret
  ret.state = createApp(objCCallback, addr ret[])
  ret.pollInterval = pollInterval
  ret.pollTimeout = pollTimeout
  ret.onNotificationClick = onNotificationClick
  return ret

proc doPoll(center: NotificationCenter) {.async.} =
  while center.polling:
    poll(center.state, center.pollTimeout / 1000)
    await sleepAsync(center.pollInterval.int)

proc show*(center: NotificationCenter,
           title, message: string,
           actionButtons: seq[tuple[ident: string, title: string]],
           subtitle = "",
           actionButtonTitle = "",
           otherButtonTitle = "",
           hasReplyButton = false) {.async.} =
  if not center.polling:
    center.polling = true
    center.pollFut = doPoll(center)

  var additionalButtons: seq[AdditionalButton] = @[]
  for i in actionButtons:
    additionalButtons.add(AdditionalButton(title: i.title, identifier: i.ident))

  var errorCode = 0.cint
  let ret = showNotification(title, subtitle, message, actionButtonTitle,
      otherButtonTitle, hasReplyButton,
      if additionalButtons.len > 0: addr additionalButtons[0] else: nil,
      additionalButtons.len.cint, addr errorCode)
  if ret != 0:
    case errorCode
    of 0:
      raise newException(NotificationError, "Error code reports no error.")
    of 1:
      raise newException(NotificationError,
          "Default Notification Center is nil. Have you specified the value" &
          " for the CFBundleIdentifier key in Info.plist?")
    else:
      raise newException(NotificationError, "Unknown error code.")

  # Only hacks available to see when the notification actually disappears.
  # Let's just assume for now (TODO) that it takes 5 seconds for a notification
  # to disappear on its own.
  await center.pollFut or sleepAsync(5000)

proc show*(center: NotificationCenter,
           title, message: string,
           subtitle = "",
           actionButtonTitle = "",
           otherButtonTitle = "",
           hasReplyButton = false): Future[void] =
  show(center, title, message, @[], subtitle, actionButtonTitle,
      otherButtonTitle, hasReplyButton)

when isMainModule:
  proc onNotificationClick(info: ClickInfo) =
    echo("Notification clicked: ", info)

  var center = newNotificationCenter(onNotificationClick)

  waitFor center.show("Nim", "Version 1.0 has been released!",
      @[("#1", "Kill All Humans"), ("#2", "Kill John Locke"),
       ("#3", "Join Star Helix")],
      actionButtonTitle="Action", otherButtonTitle="Other")
  echo("First done!")

  waitFor center.show("Sept 1st", "Hello World")
  echo("Second done!")

  waitFor center.show("Please reply", "What's your name, stranger?",
    hasReplyButton = true)
  echo("Third done!")

File: notifications/macosx.m
// Copyright (C) Dominik Picheta. All rights reserved.
// MIT License. Look at license.txt for more info.

#import <Foundation/Foundation.h>
#import <Cocoa/Cocoa.h>

typedef struct AdditionalButton {
  const char* title;
  const char* identifier;
} AdditionalButton;

typedef struct ActivationInfo {
  int activationType;
  const char* selectedActionTitle;
  const char* selectedActionIdentifier;
  const char* reply;
} ActivationInfo;

typedef void (*NotificationCallback)(ActivationInfo notification, void* data);

typedef struct NotificationState {
  NSApplication *app;
  NotificationCallback onNotificationClick;
  void* data;
} NotificationState;

// AppDelegate

@interface AppDelegate : NSObject <NSApplicationDelegate, NSUserNotificationCenterDelegate>
  @property NotificationState notificationState;
@end

@implementation AppDelegate


- (void)dealloc {
  [super dealloc];
}

- (void)applicationDidFinishLaunching:(NSNotification *)aNotification {
  [[NSUserNotificationCenter defaultUserNotificationCenter] setDelegate:self];
}

- (BOOL)userNotificationCenter:(NSUserNotificationCenter *)center
        shouldPresentNotification:(NSUserNotification *)notification {
    return YES;
}

- (void) userNotificationCenter:(NSUserNotificationCenter *)center
         didActivateNotification:(NSUserNotification *)notification {
    ActivationInfo info;
    memset(&info, 0, sizeof(ActivationInfo));
    info.activationType = (int)notification.activationType;
    if (notification.additionalActivationAction != nil) {
      info.selectedActionTitle =
        strdup([notification.additionalActivationAction.title UTF8String]);
      info.selectedActionIdentifier =
        strdup([notification.additionalActivationAction.identifier UTF8String]);
    }
    if (notification.response != nil) {
      info.reply = strdup([notification.response.string UTF8String]);
    }
    self.notificationState.onNotificationClick(info, self.notificationState.data);
}


- (void)userNotificationCenter:(NSUserNotificationCenter *)center
        didDeliverNotification:(NSUserNotification *)userNotification {
    //NSLog(@"Delivered notification");
}

@end

// Functions to be wrapped in Nim.

void freeActivationInfo(ActivationInfo info) {
  free(info.selectedActionTitle);
  free(info.selectedActionIdentifier);
  free(info.reply);
}

NotificationState createApp(NotificationCallback onNotificationClick,
    void* data) {
  NotificationState result;
  AppDelegate *delegate = [[AppDelegate alloc] init];

  NSApplication* app = [NSApplication sharedApplication];
  result.app = app;
  result.onNotificationClick = onNotificationClick;
  result.data = data;
  delegate.notificationState = result;
  [app setDelegate:delegate];
  [NSApp finishLaunching];
  return result;
}

/*
  Checks for any NSApplication events. Timeout is in seconds, but values
  between 0 and 1 are allowed so you can specify it in miliseconds.
*/
void poll(NotificationState notificationState, float timeout) {
  NSEvent *event =
    [notificationState.app
        nextEventMatchingMask:NSAnyEventMask
        untilDate:[NSDate dateWithTimeIntervalSinceNow:timeout]
        inMode:NSDefaultRunLoopMode
        dequeue:YES];
  [notificationState.app sendEvent:event];
  [notificationState.app updateWindows];
}

/*
  Shows a new notification with the specified options.

  Returns 0 on success. 1 otherwise.

  On failure the ``errorCode`` is populated with an error code. It is also
  populated when there is no error (with 0). The error codes are:

  0 - No error.
  1 - Default notification center is nil. Make sure `CFBundleIdentifier` is
      defined in the Info.plist file.
*/

int showNotification(const char* title, const char* subtitle,
                     const char* message, const char* actionButtonTitle,
                     const char* otherButtonTitle, bool hasReplyButton,
                     const AdditionalButton* additionalButtons,
                     int additionalButtonsSize,
                     int* errorCode) {
  // This proved useful for debugging this http://stackoverflow.com/a/20226193/492186

  // Allocate a new NSString based on the char* passed from Nim.
  NSString *nsTitle = [[NSString alloc] initWithUTF8String:title];
  NSString *nsSubtitle = [[NSString alloc] initWithUTF8String:subtitle];
  NSString *nsMsg = [[NSString alloc] initWithUTF8String:message];
  NSString *nsActionButtonTitle = [[NSString alloc] initWithUTF8String:actionButtonTitle];
  NSString *nsOtherButtonTitle = [[NSString alloc] initWithUTF8String:otherButtonTitle];

  NSUserNotificationCenter *center = [NSUserNotificationCenter defaultUserNotificationCenter];
  if (center == nil) {
    *errorCode = 1;
    return 1;
  }

  NSUserNotification *notification = [NSUserNotification new];
  notification.title = nsTitle;
  notification.subtitle = nsSubtitle;
  notification.informativeText = nsMsg;

  if (nsActionButtonTitle.length > 0 || nsOtherButtonTitle.length > 0 ||
      additionalButtonsSize > 0) {
    // Action buttons are off by default. You can force them to show though with
    // http://stackoverflow.com/a/23087567/492186
    [notification setValue:@YES forKey:@"_showsButtons"];
  }
  if (additionalButtonsSize > 0) {
    // https://github.com/indragiek/NSUserNotificationPrivate/blob/ad4fc22dd48495e4b5e8a6dc564113a8650338b9/NSUserNotificationPrivate/INDAppDelegate.m#L35
    // Force the showing of drop down arrow.
    [notification setValue:@YES forKey:@"_alwaysShowAlternateActionMenu"];

    NSMutableArray* additionalActions = [NSMutableArray array];
    for (int i = 0; i < additionalButtonsSize; i++) {
      NSString *nsBtnTitle = [[NSString alloc]
          initWithUTF8String:additionalButtons[i].title];
      NSString *nsBtnId = [[NSString alloc]
          initWithUTF8String:additionalButtons[i].identifier];
      [additionalActions addObject:
        [NSUserNotificationAction actionWithIdentifier:nsBtnId
          title:nsBtnTitle]];
    }
    notification.additionalActions = additionalActions;
  }

  if (nsActionButtonTitle.length > 0 || additionalButtonsSize > 0) {
    notification.hasActionButton = YES;
  }
  notification.hasReplyButton = hasReplyButton;
  notification.actionButtonTitle = nsActionButtonTitle;
  notification.otherButtonTitle = nsOtherButtonTitle;

  notification.soundName = NSUserNotificationDefaultSoundName;
  [center deliverNotification:notification];

  errorCode = 0;
  return 0;
}